quint/mempool/Mempool.qnt

```diff
@@ -0,0 +1,172 @@
+    pure def isValid(l: LedgerState, p: Process, t: Tx): bool = {
+      not(isCommittedFor(l, p, t))
+    }
+
+}
+
+module Mempool {
+
+    import Spells.* from "Spells"
+    import System.* from "System"
+    import Ledger.* from "Ledger"
+    import ABCI.*
+
+    type MempoolState = {
+      pool: Process -> Set[Tx],
+    }
+
+    pure def txsAvailable(state: MempoolState, p: Process): Set[Tx] = {
+      state.pool.get(p)
+    }
+
+    pure def poolSize(state: MempoolState, p: Process): int = {
+      state.pool.get(p).size()
+    }
+
+    pure def reap(state: MempoolState, p: Process, max: int): Set[Tx] = {
+      setSubsetOfAtMost(state.pool.get(p), max)
+    }
+
+    pure def poolOf(state: MempoolState, p: Process): Set[Tx] = {
+      state.pool.get(p)
+    }
+
+    pure def newMemPool(P: Set[Process]): MempoolState = {
+      pool: P.mapBy(p => Set()),
+    }
+
+    pure def add(state: MempoolState, p: Process, t: Tx): MempoolState = { // aka. checkTx
+      state.with("pool", state.pool.set(p, state.pool.get(p).union(Set(t))))
+    }
+
+    pure def clear(state: MempoolState, p: Process): MempoolState = {
+      state.with("pool", state.pool.set(p, Set()))
+    }
+
+    pure def remove(state: MempoolState, p: Process, t: Tx): MempoolState = {
+      state.with("pool", state.pool.set(p, state.pool.get(p).exclude(Set(t))))
+    }
+
+    // remove committed and invalid txs from the mempool
+    pure def update(state: MempoolState, p: Process, l: LedgerState, ctxs: Set[Tx]): MempoolState = {
+          state.with("pool", state.pool.set(p, state.pool.get(p).exclude(ctxs).filter(t => isValid(l, p, t))))
+    }
+
+}
+
+module MempoolTests {
+
+    import Spells.* from "Spells"
+    import System.* from "System"
+    import Ledger.* from "Ledger"
+    import Mempool.*
+    import ABCI.*
+
+    pure val PROCESSES=Set("alice", "bob")
+    pure val TXS=Set("T0", "T1", "T2")
+
+    var ledger: LedgerState
+    var mempool: MempoolState
+    var hmempool: Process -> Set[Tx] // history variable 
+
+    val print = {ledger: ledger, mempool: mempool, hmempool: hmempool}
+
+    action init : bool = all {
+      val nledger = newLedger(PROCESSES)
+      all {
+        ledger' = nledger,
+        mempool' = newMemPool(PROCESSES),
+	hmempool' = PROCESSES.mapBy(p => Set())
+      }
+    }
+
+    action doClient(p: Process, t: Tx): bool = all {
+      all {
+        require(isValid(ledger, p, t)),
+	ledger' = ledger,
+        mempool' = add(mempool, p, t),
+	hmempool' = hmempool.set(p, hmempool.get(p).union(Set(t)))
+      }
+    }
+
+    action doSubmit(p: Process): bool = all {
+      val txs = reap(mempool, p, 1)
+      all {
+        require(txs.forall(t => isValid(ledger, p, t)) and maySubmit(ledger, p, txs)),
+	ledger' = submit(ledger, p, txs),
+      	mempool' = mempool,
+	hmempool' = hmempool
+      }
+    }
+
+    action doCommitThenUpdate(p: Process): bool = all {
+      val nledger = if (mayCommit(ledger, p)) commit(ledger, p) else ledger
```

Comment:
Below is required that `mayCommit(ledger, p)` is true, so `nledger` will always be `commit(ledger, p)`.

Reply:
Indeed. This comes from the fact that it is not possible to have x' = f(y'), with x and y two state variables. As a consequence, I wrote things this way.  In the same vein, the "if" is also not necessary for the computation of the new mempool value. I have removed both "if".